# Melbourn House Price Predition Model
- created by :
    - **Student of COEP technological University, PUNE**
        1. 732392024 -- AKSHAY GIDDE


### Importing Necessory Libraries

In [749]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline      
#Magic command -- no need to use plt.show() for every graph


#for ignoring warning.
import warnings
# Ignore specific UserWarning
warnings.filterwarnings("ignore", message="The figure layout has changed to tight")

### Read csv data file and storing it in df

In [750]:
df = pd.read_csv('Dataset.csv')

- Let's check the shape of our datasets ( number of rows, number of columns/features )

In [751]:
df.shape

(13580, 21)

# Understanding the data

In [752]:
#returning random 5 rows to get to know the data
df

Suburb           Address  Rooms Type      Price Method  \
0         Abbotsford      85 Turner St      2    h  1480000.0      S   
1         Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
2         Abbotsford      5 Charles St      3    h  1465000.0     SP   
3         Abbotsford  40 Federation La      3    h   850000.0     PI   
4         Abbotsford       55a Park St      4    h  1600000.0     VB   
...              ...               ...    ...  ...        ...    ...   
13575  Wheelers Hill      12 Strada Cr      4    h  1245000.0      S   
13576   Williamstown     77 Merrett Dr      3    h  1031000.0     SP   
13577   Williamstown       83 Power St      3    h  1170000.0      S   
13578   Williamstown      96 Verdon St      4    h  2500000.0     PI   
13579     Yarraville        6 Agnes St      4    h  1285000.0     SP   

        SellerG        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  \
0        Biggin   3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0   
1        Biggin   4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0   
2        Biggin   4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0   
3        Biggin   4/03/2017       2.5    3067.0  ...       2.0  1.0      94.0   
4        Nelson   4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0   
...         ...         ...       ...       ...  ...       ...  ...       ...   
13575     Barry  26/08/2017      16.7    3150.0  ...       2.0  2.0     652.0   
13576  Williams  26/08/2017       6.8    3016.0  ...       2.0  2.0     333.0   
13577     Raine  26/08/2017       6.8    3016.0  ...       2.0  4.0     436.0   
13578   Sweeney  26/08/2017       6.8    3016.0  ...       1.0  5.0     866.0   
13579   Village  26/08/2017       6.3    3013.0  ...       1.0  1.0     362.0   

       BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0               NaN        NaN        Yarra -37.79960   144.99840   
1              79.0     1900.0        Yarra -37.80790   144.99340   
2             150.0     1900.0        Yarra -37.80930   144.99440   
3               NaN        NaN        Yarra -37.79690   144.99690   
4             142.0     2014.0        Yarra -37.80720   144.99410   
...             ...        ...          ...       ...         ...   
13575           NaN     1981.0          NaN -37.90562   145.16761   
13576         133.0     1995.0          NaN -37.85927   144.87904   
13577           NaN     1997.0          NaN -37.85274   144.88738   
13578         157.0     1920.0          NaN -37.85908   144.89299   
13579         112.0     1920.0          NaN -37.81188   144.88449   

                       Regionname Propertycount  
0           Northern Metropolitan        4019.0  
1           Northern Metropolitan        4019.0  
2           Northern Metropolitan        4019.0  
3           Northern Metropolitan        4019.0  
4           Northern Metropolitan        4019.0  
...                           ...           ...  
13575  South-Eastern Metropolitan        7392.0  
13576        Western Metropolitan        6380.0  
13577        Western Metropolitan        6380.0  
13578        Western Metropolitan        6380.0  
13579        Western Metropolitan        6543.0  

[13580 rows x 21 columns]

### Understanding each column
1. Rooms: Number of rooms

2. Price: Price in dollars

3. Method: S - property sold; SP - property sold prior; PI - property passed in; VB - vendor bid; SA - sold after auction.

4. Type: h - house,cottage,villa, semi,terrace; 
         u - unit, duplex; 
         t - townhouse; dev site - development site.

5. SellerG: Real Estate Agent

6. Date: Date sold

7. Distance: Distance from CBD

8. Regionname: General Region (West, North West, North, North east …etc)

9. Propertycount: Number of properties that exist in the suburb.

10. Bedroom2 : Scraped # of Bedrooms (from different source)

11. Bathroom: Number of Bathrooms

12. Car: Number of carspots

13. Landsize: Land Size

14. BuildingArea: Building Size

15. CouncilArea: Governing council for the area

# EDA - Exploratory Data Analysis

- cheecking 1st 5 rows of df

In [753]:
df.head()

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
1  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
2  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   
3  4/03/2017       2.5    3067.0  ...       2.0  1.0      94.0           NaN   
4  4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0         142.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   
3        NaN        Yarra  -37.7969    144.9969  Northern Metropolitan   
4     2014.0        Yarra  -37.8072    144.9941  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

- lets check the last 5 rows of df

In [754]:
df.tail()

Suburb        Address  Rooms Type      Price Method   SellerG  \
13575  Wheelers Hill   12 Strada Cr      4    h  1245000.0      S     Barry   
13576   Williamstown  77 Merrett Dr      3    h  1031000.0     SP  Williams   
13577   Williamstown    83 Power St      3    h  1170000.0      S     Raine   
13578   Williamstown   96 Verdon St      4    h  2500000.0     PI   Sweeney   
13579     Yarraville     6 Agnes St      4    h  1285000.0     SP   Village   

             Date  Distance  Postcode  ...  Bathroom  Car  Landsize  \
13575  26/08/2017      16.7    3150.0  ...       2.0  2.0     652.0   
13576  26/08/2017       6.8    3016.0  ...       2.0  2.0     333.0   
13577  26/08/2017       6.8    3016.0  ...       2.0  4.0     436.0   
13578  26/08/2017       6.8    3016.0  ...       1.0  5.0     866.0   
13579  26/08/2017       6.3    3013.0  ...       1.0  1.0     362.0   

       BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
13575           NaN     1981.0          NaN -37.90562   145.16761   
13576         133.0     1995.0          NaN -37.85927   144.87904   
13577           NaN     1997.0          NaN -37.85274   144.88738   
13578         157.0     1920.0          NaN -37.85908   144.89299   
13579         112.0     1920.0          NaN -37.81188   144.88449   

                       Regionname Propertycount  
13575  South-Eastern Metropolitan        7392.0  
13576        Western Metropolitan        6380.0  
13577        Western Metropolitan        6380.0  
13578        Western Metropolitan        6380.0  
13579        Western Metropolitan        6543.0  

[5 rows x 21 columns]

- Let's check the random ten number of data samples, Every time it will print the random five sample of records from original datasets. So we can easly understand the behaviour and what types of data type stored in particular features.

In [755]:
df.sample(10)

Suburb                Address  Rooms Type      Price Method  \
7865      Footscray      8/11 Nicholson St      2    u   392500.0     PI   
11875        Albion        2/20 Talmage St      2    t   340000.0      S   
12155      Seaholme      1/27 Noordenne Av      1    u   312000.0      S   
10960     Caulfield          22 Masters St      2    h  1070000.0      S   
9202   Keilor Downs          58 Tarella Dr      3    h   580000.0     SP   
1640     Canterbury  3/13 Wattle Valley Rd      2    u   830000.0      S   
12918  Broadmeadows         48 Electric St      3    h   637000.0      S   
2365       Essendon       4/23 Fletcher St      2    u   456000.0      S   
3438    Keilor East         14 Hawthorn Ct      3    h   590000.0      S   
5591    South Yarra          7/32 Marne St      2    u   695000.0     PI   

         SellerG        Date  Distance  Postcode  ...  Bathroom  Car  \
7865     Sweeney   6/05/2017       6.4    3011.0  ...       1.0  1.0   
11875      Bells  29/07/2017      10.5    3020.0  ...       1.0  1.0   
12155         RT  29/07/2017      11.0    3018.0  ...       1.0  0.0   
10960       Gary  12/08/2017       8.7    3162.0  ...       1.0  0.0   
9202         Ace   3/06/2017      15.5    3038.0  ...       1.0  4.0   
1640     Century   4/06/2016       9.0    3126.0  ...       1.0  1.0   
12918  Stockdale  19/08/2017      14.0    3047.0  ...       1.0  5.0   
2365        Brad   8/10/2016       8.0    3040.0  ...       1.0  0.0   
3438      Nelson   3/12/2016      12.8    3033.0  ...       1.0  2.0   
5591         Kay   4/03/2017       3.3    3141.0  ...       1.0  0.0   

       Landsize  BuildingArea  YearBuilt    CouncilArea Lattitude  Longtitude  \
7865     1783.0           NaN     1960.0    Maribyrnong -37.79490   144.90070   
11875     100.0           NaN        NaN       Brimbank -37.77956   144.82519   
12155       0.0           NaN        NaN    Hobsons Bay -37.86007   144.84578   
10960     261.0           NaN        NaN      Glen Eira -37.88569   145.02496   
9202      677.0           NaN        NaN       Brimbank -37.72188   144.81489   
1640      110.0          83.0     1960.0     Boroondara -37.82640   145.08240   
12918     670.0           NaN     1980.0            NaN -37.69215   144.91654   
2365        0.0           NaN        NaN  Moonee Valley -37.75570   144.92220   
3438      610.0           NaN     1974.0  Moonee Valley -37.74550   144.84840   
5591        0.0          79.0     1930.0      Melbourne -37.83610   144.98420   

                  Regionname Propertycount  
7865    Western Metropolitan        7570.0  
11875   Western Metropolitan        2185.0  
12155   Western Metropolitan         852.0  
10960  Southern Metropolitan        2379.0  
9202    Western Metropolitan        3656.0  
1640   Southern Metropolitan        3265.0  
12918  Northern Metropolitan        4294.0  
2365    Western Metropolitan        9264.0  
3438    Western Metropolitan        5629.0  
5591   Southern Metropolitan       14887.0  

[10 rows x 21 columns]

- lets check column wise information of our dataset

In [756]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         13580 non-null  object 
 1   Address        13580 non-null  object 
 2   Rooms          13580 non-null  int64  
 3   Type           13580 non-null  object 
 4   Price          13580 non-null  float64
 5   Method         13580 non-null  object 
 6   SellerG        13580 non-null  object 
 7   Date           13580 non-null  object 
 8   Distance       13580 non-null  float64
 9   Postcode       13580 non-null  float64
 10  Bedroom2       13580 non-null  float64
 11  Bathroom       13580 non-null  float64
 12  Car            13518 non-null  float64
 13  Landsize       13580 non-null  float64
 14  BuildingArea   7130 non-null   float64
 15  YearBuilt      8205 non-null   float64
 16  CouncilArea    12211 non-null  object 
 17  Lattitude      13580 non-null  float64
 18  Longti

#### From the above output we can see that :
1. There are 13580 unique rows
2. 21 column/features in the dataset i.e.[0 to 20]
3. Datatypes are of 3 types : object, float, int64
4. There 4 column which are having null values in it.
    - column with null values [ BuildingArea , YearBuilt , CouncilArea , Car ]
5. Memory usage is 2.2+ MB which is negligible and easily handled by machine.

### Handling Null values
- lets check the number of null in each column

In [757]:
df.isnull().sum().sort_values(ascending = False)

BuildingArea     6450
YearBuilt        5375
CouncilArea      1369
Car                62
Suburb              0
Bathroom            0
Regionname          0
Longtitude          0
Lattitude           0
Landsize            0
Bedroom2            0
Address             0
Postcode            0
Distance            0
Date                0
SellerG             0
Method              0
Price               0
Type                0
Rooms               0
Propertycount       0
dtype: int64

- Here, We convert the number of missing values into percentages. So, we can easly understand to how many percentage of missing values available.

In [758]:
# Gives the % of NULL values from the total NULL values down the column
((df.isnull().sum()*100)/df.isnull().count()).sort_values(ascending = False).head(5)

BuildingArea    47.496318
YearBuilt       39.580265
CouncilArea     10.081001
Car              0.456554
Suburb           0.000000
dtype: float64

- Here we can see [ CouncilArea , Car ] are the two columns which have null values percentage below or around 10 % so we can fill those values using statistical methods. Like mean, mode, median

- Here we can see [ BuildingArea, YearBuilt ] are the two colums where the NULL value percentage is very high which means close to half of the data is null. So, filling this makes no sense and it will also divert the model AND it will make wrong feeding to the model.
- So, we will drop thosee 2 columns with high percentage of NULL values.

In [759]:
# dropping  [ BuildingArea, YearBuilt ] columns
df.drop(columns = ['BuildingArea','YearBuilt',],inplace = True)

**filling car column with the median of the car values**
- Here mode = median != mean
- mean is prone to outliers so we are filling medain

In [760]:
#Filling car null with median
df['Car'].fillna(df['Car'].median(),inplace = True)

/var/folders/qr/knjrtlb17lb7d0mvwrpx6p_40000gn/T/ipykernel_29988/476660682.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Car'].fillna(df['Car'].median(),inplace = True)


In [761]:
# Filling CouncilArea's NULL with its mode (most frequent value)
df['CouncilArea'].fillna(df['CouncilArea'].mode()[0],inplace = True)

/var/folders/qr/knjrtlb17lb7d0mvwrpx6p_40000gn/T/ipykernel_29988/2721604904.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['CouncilArea'].fillna(df['CouncilArea'].mode()[0],inplace = True)


In [762]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 19 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         13580 non-null  object 
 1   Address        13580 non-null  object 
 2   Rooms          13580 non-null  int64  
 3   Type           13580 non-null  object 
 4   Price          13580 non-null  float64
 5   Method         13580 non-null  object 
 6   SellerG        13580 non-null  object 
 7   Date           13580 non-null  object 
 8   Distance       13580 non-null  float64
 9   Postcode       13580 non-null  float64
 10  Bedroom2       13580 non-null  float64
 11  Bathroom       13580 non-null  float64
 12  Car            13580 non-null  float64
 13  Landsize       13580 non-null  float64
 14  CouncilArea    13580 non-null  object 
 15  Lattitude      13580 non-null  float64
 16  Longtitude     13580 non-null  float64
 17  Regionname     13580 non-null  object 
 18  Proper

Null values has been taken care of!!!!

## Checking the distribution of target data ('Price')

In [763]:
#plotting a distribution graph of a 'Price'
plt.figure(figsize = (8,3))
sns.histplot(df['Price'])

<Axes: xlabel='Price', ylabel='Count'>

- here we clearly can see the data is not normally distributed
- Bell curve is clearly skewed towards right hand side
- **it means there are outliers in upper range i.e (4th quartile)**
- We have to take care of this

## Handling Ouliers
- taking our target data price and will handle outliers on the basis of 'Price'
- There are sevral methods of Outlier Handling Like:
    1. Z-Score method
    2. IQR method (Interquartile range)
    3. Percentile method

#### Making suitable plot to identify outliers in price

In [764]:
plt.figure(figsize=(4,5))
plt.boxplot(df['Price'])
plt.title('Box Plot With Outliers')
plt.tight_layout()

- So, after ploting the boxplot for the Price column, we can see that there are some outliers present in the price column and from this we can assume that, there are outliers beyond value ~~~ 2.2

### 1. Z-Score Method

- so to handle the outlier and get the exact value level of lower and upper limit of the price we will use Z-scoe technique (method)
- We will take values upto 3 standard deviation from mean on either side to get upper and lower limit
    - For lower_limit add 3 std.dev. in the mean
    - For lower_limit less 3 std.dev. from mean

In [765]:
print('Price mean =',round(df['Price'].mean(),2))
print('Standard dev =',round(df['Price'].std(),2),'\n')

upper_limit = df['Price'].mean() + 3 * df['Price'].std()
lower_limit = df['Price'].mean() - 3 * df['Price'].std()


print(f'upper limit: {upper_limit}')
print(f'lower limit: {lower_limit}\n')

#Lets Find outliers based on upper and lower limits 
print('total number of outliers =', len(df[(df['Price'] > upper_limit) | (df['Price'] < lower_limit)]),)

# dropping the assigned ouliers
df.drop(index=df[(df['Price'] > upper_limit) | (df['Price'] < lower_limit)].index,inplace = True)

print(f'Data rows without outlier = {len(df)}\n')


Price mean = 1075684.08
Standard dev = 639310.72 

upper limit: 2993616.252343139
lower limit: -842248.0934329773

total number of outliers = 232
Data rows without outlier = 13348



### 2. Inter Quartile Range Method

In [766]:
# #calculate q1 and q2
# q1 = df['Price'].quantile(0.25)
# q3 = df['Price'].quantile(0.75)

# #calculate iter quartile range
# iqr = q3-q1

# #calculating upper and lower limit
# upper_limit = ul = q3 + (1.5 * iqr)
# lower_limit = ll = q1 - (1.5 * iqr)


# print(f'Upper Limit : {ul}\nLower Limit : {ll}')

# #creating the index and dropping those rows which are outlier
# df.drop(index = df[df['Price']>= ul].index, inplace = True )
# df.drop(index = df[df['Price']<= ll].index, inplace = True )
 
# #printing the length of new df
# print('Rows count after Dropping Outliers :',len(df))

- **Plotting the boxplot again to see the inprovement ouliers**

In [767]:
plt.figure(figsize = (4,5))
plt.boxplot(df['Price'])
plt.title('Box Plot After Removing Outliers')
plt.tight_layout()

### Distribution after oulier handling
- its somewhat normal now so we can proceed

In [768]:
plt.figure(figsize=(8,3))
sns.histplot(df['Price'])

<Axes: xlabel='Price', ylabel='Count'>

#### Outliers has been taken care of

### Make categoriacal and numeric features (separate)

In [769]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13348 entries, 0 to 13579
Data columns (total 19 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         13348 non-null  object 
 1   Address        13348 non-null  object 
 2   Rooms          13348 non-null  int64  
 3   Type           13348 non-null  object 
 4   Price          13348 non-null  float64
 5   Method         13348 non-null  object 
 6   SellerG        13348 non-null  object 
 7   Date           13348 non-null  object 
 8   Distance       13348 non-null  float64
 9   Postcode       13348 non-null  float64
 10  Bedroom2       13348 non-null  float64
 11  Bathroom       13348 non-null  float64
 12  Car            13348 non-null  float64
 13  Landsize       13348 non-null  float64
 14  CouncilArea    13348 non-null  object 
 15  Lattitude      13348 non-null  float64
 16  Longtitude     13348 non-null  float64
 17  Regionname     13348 non-null  object 
 18  Propertycou

In [770]:
categorical_features = df.select_dtypes('object')
numeric_features = df.select_dtypes(include = ('int64','float64'))

- Check the unique features in each of the categorical columns

In [771]:
for col in categorical_features:
    print(f'Unique values in {col} = {len(df[col].unique())}')

Unique values in Suburb = 314
Unique values in Address = 13151
Unique values in Type = 3
Unique values in Method = 5
Unique values in SellerG = 267
Unique values in Date = 58
Unique values in CouncilArea = 33
Unique values in Regionname = 8


- Each unique feature will become its own feature in the dataset, resulting in a high-dimensional feature space.
- This can make the model more complex and computationally intensive, especially if the address feature has a large number of unique values.
-  to higher number of unique values(close to 99%)in 'Address' column it will heavly impact the complexity of the overall model.
- so we have to delete such columns. (like Address, Suburb, SellerG)

In [772]:
categorical_features.drop(columns = ['Suburb','Address','SellerG','Date'],inplace = True)

- Checking the columunar uniqueness again:

In [773]:
for col in categorical_features:
    print(f'Unique values in {col} = {len(df[col].unique())}')
    

Unique values in Type = 3
Unique values in Method = 5
Unique values in CouncilArea = 33
Unique values in Regionname = 8


### Feature Modification
- **Here, We need to convert categorical values to numerical values**
- For that we use pandas inbuilt **.get_dummies** fucntion.
- **.get_dummies** do kind of one hot encoding which makes a new column for each unique feature and assign binnary to it


In [774]:
#take a look at categorical features
categorical_features

Type Method CouncilArea                  Regionname
0        h      S       Yarra       Northern Metropolitan
1        h      S       Yarra       Northern Metropolitan
2        h     SP       Yarra       Northern Metropolitan
3        h     PI       Yarra       Northern Metropolitan
4        h     VB       Yarra       Northern Metropolitan
...    ...    ...         ...                         ...
13575    h      S    Moreland  South-Eastern Metropolitan
13576    h     SP    Moreland        Western Metropolitan
13577    h      S    Moreland        Western Metropolitan
13578    h     PI    Moreland        Western Metropolitan
13579    h     SP    Moreland        Western Metropolitan

[13348 rows x 4 columns]

- Encoding the actual data by using get dummies which is pandas default encoder
- creating new encoded data columns attaching those new columns to df and dropping original once

In [775]:
categorical_features = categorical_features.join(pd.get_dummies(categorical_features)).drop(columns = categorical_features)

In [776]:
categorical_features.shape

(13348, 49)

In [777]:
numeric_features.shape

(13348, 11)

In [778]:
categorical_features

Type_h  Type_t  Type_u  Method_PI  Method_S  Method_SA  Method_SP  \
0        True   False   False      False      True      False      False   
1        True   False   False      False      True      False      False   
2        True   False   False      False     False      False       True   
3        True   False   False       True     False      False      False   
4        True   False   False      False     False      False      False   
...       ...     ...     ...        ...       ...        ...        ...   
13575    True   False   False      False      True      False      False   
13576    True   False   False      False     False      False       True   
13577    True   False   False      False      True      False      False   
13578    True   False   False       True     False      False      False   
13579    True   False   False      False     False      False       True   

       Method_VB  CouncilArea_Banyule  CouncilArea_Bayside  ...  \
0          False                False                False  ...   
1          False                False                False  ...   
2          False                False                False  ...   
3          False                False                False  ...   
4           True                False                False  ...   
...          ...                  ...                  ...  ...   
13575      False                False                False  ...   
13576      False                False                False  ...   
13577      False                False                False  ...   
13578      False                False                False  ...   
13579      False                False                False  ...   

       CouncilArea_Yarra  CouncilArea_Yarra Ranges  \
0                   True                     False   
1                   True                     False   
2                   True                     False   
3                   True                     False   
4                   True                     False   
...                  ...                       ...   
13575              False                     False   
13576              False                     False   
13577              False                     False   
13578              False                     False   
13579              False                     False   

       Regionname_Eastern Metropolitan  Regionname_Eastern Victoria  \
0                                False                        False   
1                                False                        False   
2                                False                        False   
3                                False                        False   
4                                False                        False   
...                                ...                          ...   
13575                            False                        False   
13576                            False                        False   
13577                            False                        False   
13578                            False                        False   
13579                            False                        False   

       Regionname_Northern Metropolitan  Regionname_Northern Victoria  \
0                                  True                         False   
1                                  True                         False   
2                                  True                         False   
3                                  True                         False   
4                                  True                         False   
...                                 ...                           ...   
13575                             False                         False   
13576                             False                         False   
13577                             False                         False   
13578                             False                  

In [779]:
numeric_features

Rooms      Price  Distance  Postcode  Bedroom2  Bathroom  Car  \
0          2  1480000.0       2.5    3067.0       2.0       1.0  1.0   
1          2  1035000.0       2.5    3067.0       2.0       1.0  0.0   
2          3  1465000.0       2.5    3067.0       3.0       2.0  0.0   
3          3   850000.0       2.5    3067.0       3.0       2.0  1.0   
4          4  1600000.0       2.5    3067.0       3.0       1.0  2.0   
...      ...        ...       ...       ...       ...       ...  ...   
13575      4  1245000.0      16.7    3150.0       4.0       2.0  2.0   
13576      3  1031000.0       6.8    3016.0       3.0       2.0  2.0   
13577      3  1170000.0       6.8    3016.0       3.0       2.0  4.0   
13578      4  2500000.0       6.8    3016.0       4.0       1.0  5.0   
13579      4  1285000.0       6.3    3013.0       4.0       1.0  1.0   

       Landsize  Lattitude  Longtitude  Propertycount  
0         202.0  -37.79960   144.99840         4019.0  
1         156.0  -37.80790   144.99340         4019.0  
2         134.0  -37.80930   144.99440         4019.0  
3          94.0  -37.79690   144.99690         4019.0  
4         120.0  -37.80720   144.99410         4019.0  
...         ...        ...         ...            ...  
13575     652.0  -37.90562   145.16761         7392.0  
13576     333.0  -37.85927   144.87904         6380.0  
13577     436.0  -37.85274   144.88738         6380.0  
13578     866.0  -37.85908   144.89299         6380.0  
13579     362.0  -37.81188   144.88449         6543.0  

[13348 rows x 11 columns]

#### Joining categorical_featues and numeric_features and storing them in one data (dataframe)

In [780]:
data = categorical_features.join(numeric_features)

### Splitting data into training and target variables

- creating a list which has heading of each of the columns 

In [781]:
features = list(data)

In [782]:
features.remove('Price')

- 'Price' is our target variable so, removing it from the list of features so that list of features can go to x whcih is training features

In [783]:
training_features = x = data[features]

In [784]:
target_features = y = data['Price']

### Splitting data into Train Test Split

In [785]:
# importing train test split
from sklearn.model_selection import train_test_split

**Splites the main data**

- split data into training and validation data, for both features and target. 
- The split is based on a random number generator.
- Supplying a numeric value to the random_state argument guarantees we get the same split even run this script.
- split test size will be of test = 25%

In [786]:
Xtrain, Xtest, Ytrain, Ytest = train_test_split(x, y, test_size = 0.15,random_state=4)

In [787]:
print("Total size: ", x.shape)
print("Train size: ", Xtrain.shape, Ytrain.shape)
print("Test size: ", Xtest.shape, Ytest.shape)

Total size:  (13348, 59)
Train size:  (11345, 59) (11345,)
Test size:  (2003, 59) (2003,)


---

### Model Building
- As per the guidline of the project we will use SVR
- Although SVR is a classification based regressor

## SVR

In [788]:
# SVR model is storing in model variable
model = SVR()

In [789]:
# it will make sure each columns will have names assigned
model.feature_names_in_ = features

In [790]:
# Fit/Feed the train data to the model
model.fit(Xtrain,Ytrain)

SVR()

- To generates predictions for the test data based on the trained model, which are stored in the Ypred variable.

In [791]:
Ypred = model.predict(Xtest)

### Accuracy testing
- this is a prediction algorithm so confusion matrix can't be used.
- so we will be using mean absolute error and R^2 error to evalute our model.


In [792]:
# Importing necessory library for testing 
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import r2_score

In [793]:
# mean squared error
mae = mean_absolute_error(Ytest,Ypred)
print(f'mean absolute error = {mae}\n')


# R^2 score
r2 = r2_score(Ytest, Ypred)
print("R^2 Score:", r2)

mean absolute error = 411047.30953821767

R^2 Score: -0.08673428124577898


### **SVR is failing in this type of algorithm so we will use another regressor algorithms to predict the house prices** 

---

## 1. Decision Tree Regressor

In [794]:
from sklearn.tree import DecisionTreeRegressor

In [795]:
# #to get the best random state for model
# best_score = float('inf')  # Initialize with a high value
# best_random_state = None


# for random_state in range(100):  # Try different random states
#     model = DecisionTreeRegressor(random_state=random_state)
#     model.fit(Xtrain, Ytrain)
#     Ypred = model.predict(Xtest)
#     score = mean_absolute_error(Ytest, Ypred)  # Evaluate using mean squared error
#     if score < best_score:
#         best_score = score
#         best_random_state = random_state

# print("Best Random State:", best_random_state)
# print("Best Mean Squared Error:", best_score)

In [796]:
dtm = DecisionTreeRegressor()

In [797]:
dtm.fit(Xtrain, Ytrain)

DecisionTreeRegressor()

In [798]:
dtm_Ypred = dtm.predict(Xtest)

In [799]:
# mean squared error
mae = mean_absolute_error(Ytest,dtm_Ypred)
print(f'mean absolute error = {mae}\n')


# R^2 score
r2 = r2_score(Ytest, dtm_Ypred)
print('Accuracy = R^2 = ',round(r2*100,2),'%')

mean absolute error = 201600.5521717424

Accuracy = R^2 =  69.12 %


---


## 2. Linear Regression Model

In [800]:
 from sklearn.linear_model import LinearRegression

In [801]:
lrm = LinearRegression()

In [802]:
lrm.fit(Xtrain,Ytrain)

LinearRegression()

In [803]:
lrm_Ypred = lrm.predict(Xtest)

In [804]:
# mean squared error
mae = mean_absolute_error(Ytest,lrm_Ypred)
print(f'mean absolute error = {mae}\n')


# R^2 score
r2 = r2_score(Ytest, lrm_Ypred)
print('Accuracy = R^2 = ',round(r2*100,2),'%')

mean absolute error = 223098.241407546

Accuracy = R^2 =  68.11 %


---

## 3. Rigid Regression

In [805]:
from sklearn.linear_model import Ridge

In [806]:
rr = Ridge(alpha=0.01)

In [807]:
rr.fit(Xtrain, Ytrain)

Ridge(alpha=0.01)

In [808]:
rr_Ypred = rr.predict(Xtest)

In [809]:
# mean squared error
mae = mean_absolute_error(Ytest,rr_Ypred)
print(f'mean absolute error = {mae}\n')


# R^2 score
r2 = r2_score(Ytest,rr_Ypred)
print('Accuracy = R^2 = ',round(r2*100,2),'%')

mean absolute error = 223096.76214582307

Accuracy = R^2 =  68.11 %


#### Line Equation for Ridge

In [810]:
ridge_equation = f'Price = {rr.intercept_} + '
for coef,feature in zip(ls.coef_,features):
    ridge_equation += f'[({coef}){feature}] + '

ridge_equation

'Price = -100714061.6973198 + [(441241.8551085894)Type_h] + [(160833.27353339488)Type_t] + [(-57425.676874713696)Type_u] + [(-24542.716475712565)Method_PI] + [(43123.34526369247)Method_S] + [(-375.52654338482085)Method_SA] + [(1724.5712084006477)Method_SP] + [(-16702.366803034)Method_VB] + [(-89224.57379727949)CouncilArea_Banyule] + [(206517.87726314954)CouncilArea_Bayside] + [(192767.0167832455)CouncilArea_Boroondara] + [(-124564.95678416044)CouncilArea_Brimbank] + [(-222700.17623408115)CouncilArea_Cardinia] + [(-428632.26772657706)CouncilArea_Casey] + [(94919.52058208124)CouncilArea_Darebin] + [(-94711.33044812622)CouncilArea_Frankston] + [(-134092.90522135558)CouncilArea_Glen Eira] + [(-222907.82491573453)CouncilArea_Greater Dandenong] + [(17840.78908593665)CouncilArea_Hobsons Bay] + [(47607.980824695616)CouncilArea_Hume] + [(-165966.638472234)CouncilArea_Kingston] + [(-212681.47567982087)CouncilArea_Knox] + [(890643.6011985705)CouncilArea_Macedon Ranges] + [(16875.703447838572)Coun

---

## 4. Lasso Regression

In [811]:
from sklearn.linear_model import Lasso
ls = Lasso(alpha=1)
ls.fit(Xtrain, Ytrain)
ls_Ypred = ls.predict(Xtest)

/opt/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:697: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 7.090e+13, tolerance: 3.089e+11
  model = cd_fast.enet_coordinate_descent(


In [812]:
# mean squared error
mae = mean_absolute_error(Ytest,ls_Ypred)
print(f'mean absolute error = {mae}\n')


# R^2 score
r2 = r2_score(Ytest,ls_Ypred)
print('Accuracy = R^2 = ',round(r2*100,2),'%')

mean absolute error = 223084.89334231534

Accuracy = R^2 =  68.11 %


#### Line Equation for Lassso

In [813]:
lasso_equation = f"Price = {ls.intercept_}"
for coef, feature in zip(ls.coef_, features):
    lasso_equation += f" + [({coef})*{feature}]"
print(lasso_equation)

Price = -100776655.97504507 + [(441241.8551085894)*Type_h] + [(160833.27353339488)*Type_t] + [(-57425.676874713696)*Type_u] + [(-24542.716475712565)*Method_PI] + [(43123.34526369247)*Method_S] + [(-375.52654338482085)*Method_SA] + [(1724.5712084006477)*Method_SP] + [(-16702.366803034)*Method_VB] + [(-89224.57379727949)*CouncilArea_Banyule] + [(206517.87726314954)*CouncilArea_Bayside] + [(192767.0167832455)*CouncilArea_Boroondara] + [(-124564.95678416044)*CouncilArea_Brimbank] + [(-222700.17623408115)*CouncilArea_Cardinia] + [(-428632.26772657706)*CouncilArea_Casey] + [(94919.52058208124)*CouncilArea_Darebin] + [(-94711.33044812622)*CouncilArea_Frankston] + [(-134092.90522135558)*CouncilArea_Glen Eira] + [(-222907.82491573453)*CouncilArea_Greater Dandenong] + [(17840.78908593665)*CouncilArea_Hobsons Bay] + [(47607.980824695616)*CouncilArea_Hume] + [(-165966.638472234)*CouncilArea_Kingston] + [(-212681.47567982087)*CouncilArea_Knox] + [(890643.6011985705)*CouncilArea_Macedon Ranges] + [(

### Features with coefficients less than 0.01

In [814]:
# Find features with coefficients less than 0.01
zero_coef_features = [feature for coef, feature in zip(ls.coef_, features) if abs(coef) < 0.01]
print("Features with zero coefficients:", zero_coef_features)

Features with zero coefficients: ['CouncilArea_Moorabool']


According to theory of lasso regression it make the coef zero of irrelavant features which dosent contribue that much to the prediction.
- We can drop those columns in such case to reduce the curse of dimentionality and also improve the model performance.

---

- **From above observation we can see that all other algoriths are working fine and showing significat accuracy which is close to 70%**
- **here we can say SVR is not a best fit for this data prediction**

### Hist - Gradient Boosting Regressor
- **Hist - Gradient Boosting Regressor will be the best for this problem lets check accuracy by using this.**

In [815]:
from sklearn.ensemble import HistGradientBoostingRegressor

In [816]:
hgbr = HistGradientBoostingRegressor()

In [817]:
hgbr.fit(Xtrain,Ytrain)

HistGradientBoostingRegressor()

In [818]:
hgbr_Ypred = hgbr.predict(Xtest)

In [819]:
# mean squared error
mae = mean_absolute_error(Ytest,hgbr_Ypred)
print(f'mean absolute error = {mae}\n')


# R^2 score
r2 = r2_score(Ytest,hgbr_Ypred)
print('Accuracy = R^2 = ',round(r2*100,2),'%')

mean absolute error = 146931.98262349208

Accuracy = R^2 =  84.14 %


# Thanks